### PR TITLE
chore(test-data): create cart discount presets, add new `KeyReference` model

### PR DESCRIPTION
**Pull request by jaikamat**
## Summary

This PR makes 80% completion of the cart discount presets. I didn't want this PR to grow in scope, and will handle the remaining 20% of work as a follow-up PR. 

Comments inline, as it's easier to explain alongside code.

### CODE REVIEW

File: models/cart-discount/src/cart-discount-draft/presets/sample-data-fashion/employee-sale.ts [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+const employeeSale = (): TCartDiscountDraftBuilder =>
+  CartDiscountDraft.presets
+    .empty()
+    .value(
+      CartDiscountValueRelative.CartDiscountValueRelativeDraft.random()
+        .type('relative')
+        .permyriad(1500)
+    )
+    .cartPredicate(`customer.customerGroup.key = "${customerGroupDraft.key}"`)
+    // TODO: create `target` model
```
Comment (jaikamat) [inline models/cart-discount/src/cart-discount-draft/presets/sample-data-fashion/employee-sale.ts:23]: The remaining work is to create models for `target`, but I wanted to get this PR up as it's already growing quite large

File: models/commons/src/reference/types.ts [hunk 2]
```diff
@@ -4,12 +4,12 @@ export type TReferenceBuilder<TypeId = string> = TBuilder<TReference<TypeId>>;
...
   typeId: TypeId;
-  id: string;
-  key: string;
+  id?: string;
+  key?: string;
 }
 
 export type TExpandedReferenceObject<TypeId = string> = {
-  id: TReference<TypeId>['id'];
+  id?: TReference<TypeId>['id'];
```
Comment (jaikamat) [inline models/commons/src/reference/types.ts:12]: cc @emmenko @stephsprinkle 

This is the big change here. I made `id` and `key` optional fields, as we need a way to create an `empty` preset for reference types. The generator for this spoofs an `id`, which means if we send these to the CoCo API, we could get id/key collisions or errors. I saw two paths forward:

1. have the generator no longer spoof an `id`. This means we need to alter every implementation of `reference` to include a spoofed id. A lot of work, to be honest
2. Make both fields optional, and include an `empty` preset so we can clear existing `id`'s and add our own `key`. This is what I chose to do. 

This solution passed typecheck and all tests, so I felt it stable.

File: models/cart-discount-value-gift-line-item/src/cart-discount-value-gift-line-item-draft/presets/empty.ts [hunk 3]
```diff
@@ -0,0 +1,9 @@
+import type { TCartDiscountValueGiftLineItemDraftBuilder } from '../../types';
+import CartDiscountValueGiftLineItemBuilder from '../builder';
+
+const empty = (): TCartDiscountValueGiftLineItemDraftBuilder =>
+  CartDiscountValueGiftLineItemBuilder()
+    .distributionChannel(undefined)
+    .supplyChannel(undefined);
```
Comment (jaikamat) [inline models/cart-discount-value-gift-line-item/src/cart-discount-value-gift-line-item-draft/presets/empty.ts:7]: The gift line item generator also spoofs id's for distribution and supply channels, so we needed to be able to empty these in order to snapshot test reliably

File: models/cart-discount/src/cart-discount-draft/presets/sample-data-fashion/summer-flips.spec.ts [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+        "product": {
+          "id": undefined,
+          "key": "summer_dress",
+          "typeId": "product",
+        },
+        "supplyChannel": undefined,
+        "type": "giftLineItem",
+        "variantId": 1,
+      }
+    `);
```
Comment (jaikamat) [inline models/cart-discount/src/cart-discount-draft/presets/sample-data-fashion/summer-flips.spec.ts:20]: Now, the product `reference` has a stable `undefined` `id`, and other fields we don't care about like `distributionChannel` and `supplyChannel` are `undefined` as well

File: .changeset/cyan-seals-begin.md [hunk 5]
```diff
@@ -0,0 +1,8 @@
+---
+'@commercetools-test-data/cart-discount-value-gift-line-item': patch
+'@commercetools-test-data/cart-discount': patch
+'@commercetools-test-data/discount-code': patch
+'@commercetools-test-data/commons': patch
```
Comment (stephsprinkle) [inline .changeset/cyan-seals-begin.md:5]: These should all be `minor`

File: models/cart-discount/src/cart-discount-draft/presets/sample-data-fashion/employee-sale.ts [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+    .target(undefined)
+    .name(LocalizedString.presets.empty()['en-US']('Employee Sale'))
+    .description(LocalizedString.presets.empty()['en-US']('employee_sale'))
+    .stackingMode('Stacking')
```
Comment (stephsprinkle) [inline models/cart-discount/src/cart-discount-draft/presets/sample-data-fashion/employee-sale.ts:25]: Don't we have a constant for this value?

File: models/cart-discount/src/cart-discount-draft/presets/sample-data-fashion/luxe-spend.ts [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+        .type('absolute')
+        .money(
+          CentPrecisionMoney.random()
+            .currencyCode('EUR')
+            .centAmount(3000)
+            .fractionDigits(2)
+        )
+    )
+    .cartPredicate(
+      'totalPrice = "500.00 EUR" and customer.customerGroup.key = "luxe"'
```
Comment (stephsprinkle) [inline models/cart-discount/src/cart-discount-draft/presets/sample-data-fashion/luxe-spend.ts:21]: Customer group key should be evaluated similar to the employee sale discount

File: models/cart-discount/src/cart-discount-draft/presets/sample-data-fashion/percent-off-womens-wear.ts [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+import { LocalizedString } from '@commercetools-test-data/commons';
+import * as CartDiscountDraft from '../..';
+import { TCartDiscountDraftBuilder } from '../../../types';
+
+const percentOffWomensWear = (): TCartDiscountDraftBuilder =>
+  CartDiscountDraft.presets
+    .empty()
+    .value(
+      CartDiscountValueRelative.CartDiscountValueRelativeDraft.random()
+        .type('relative')
```
Comment (stephsprinkle) [inline models/cart-discount/src/cart-discount-draft/presets/sample-data-fashion/percent-off-womens-wear.ts]: Random thought, given the entity is `CartDiscountValueRelativeDraft` should `type('relative')` be a given? 

Care amount so low, throwing things out into the universe.

File: models/cart-discount/src/cart-discount-draft/presets/sample-data-fashion/shirts-bogo.ts [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+import * as CartDiscountDraft from '../..';
+import { TCartDiscountDraftBuilder } from '../../../types';
+
+const shirtsBogo = (): TCartDiscountDraftBuilder =>
+  CartDiscountDraft.presets
+    .empty()
+    .value(
+      CartDiscountValueRelative.CartDiscountValueRelativeDraft.random()
+        .type('relative')
+        .permyriad(1000)
```
Comment (stephsprinkle) [inline models/cart-discount/src/cart-discount-draft/presets/sample-data-fashion/shirts-bogo.ts]:
```suggestion
        .permyriad(10000)
```

File: models/commons/src/reference/types.ts (hunk 2 above)
Comment (stephsprinkle) [reply models/commons/src/reference/types.ts:12]: They other option is to create a KeyReference model and leave this one be just `id`.

**Review by stephsprinkle — COMMENTED** [review]
A couple of things - looking good! It's nice to see all the hard work coming together.

File: models/cart-discount/src/cart-discount-draft/presets/sample-data-fashion/employee-sale.ts (hunk 6 above)
Comment (jaikamat) [reply models/cart-discount/src/cart-discount-draft/presets/sample-data-fashion/employee-sale.ts]: Yep! Fixed in 863e4ba717ae8a5363ca6f47484e12aec4a8953c

File: models/cart-discount/src/cart-discount-draft/presets/sample-data-fashion/luxe-spend.ts (hunk 7 above)
Comment (jaikamat) [reply models/cart-discount/src/cart-discount-draft/presets/sample-data-fashion/luxe-spend.ts]: Nice catch, eagle eyes 👁️ 🦅 

da0a5dbbb44c6484ef21ff7554155ada5e2bda52

File: models/cart-discount/src/cart-discount-draft/presets/sample-data-fashion/shirts-bogo.ts (hunk 9 above)
Comment (jaikamat) [reply models/cart-discount/src/cart-discount-draft/presets/sample-data-fashion/shirts-bogo.ts]: Thanks! 2d4cd1b3a18f7044a64892ffc9430d81df2298c7

File: models/cart-discount/src/cart-discount-draft/presets/sample-data-fashion/percent-off-womens-wear.ts (hunk 8 above)
Comment (jaikamat) [reply models/cart-discount/src/cart-discount-draft/presets/sample-data-fashion/percent-off-womens-wear.ts]: You're not wrong 😅  017343388571b114ad5d5b557254cb091cf9bf13

File: models/commons/src/reference/types.ts (hunk 2 above)
Comment (jaikamat) [reply models/commons/src/reference/types.ts:12]: Actually nvm, rather deal with it here. Don't want the edits to `Reference` merged in

**Comment by jaikamat** [comment]
@stephsprinkle Here's the commit for `KeyReference`: 8ca26c37715d7a74ada3696eae30b13074408d21
Here's the commit for using said `KeyReference`: 7a7d0da2f88d1fe5514a83d0eae2204281e0a154

File: models/commons/src/reference/presets/empty.ts [hunk 10]
```diff
@@ -0,0 +1,6 @@
+import Reference from '../builder';
+import type { TReferenceBuilder } from '../types';
+
+const empty = (): TReferenceBuilder => Reference().id(undefined).key(undefined);
```
Comment (stephsprinkle) [inline models/commons/src/reference/presets/empty.ts:4]: Should this be just `id` now?
```suggestion
const empty = (): TReferenceBuilder => Reference().id(undefined);
```

File: models/commons/src/reference/types.ts (hunk 2 above)
Comment (stephsprinkle) [reply models/commons/src/reference/types.ts:12]: Can the optional stuff be removed now that we have `KeyReference`?

Comment (jaikamat) [reply models/commons/src/reference/types.ts]: Thanks for the reminder, goldfish memory over here

8a5b631dd1465de561e040ec9d27c1128f9c79ce

**Comment by jaikamat** [comment]
@stephsprinkle axed the preset (no longer needed) in 5957de8ca5c311b7f1bad50e46fa6fad13a5179d, missed that one

File: .changeset/cyan-seals-begin.md [hunk 11]
```diff
@@ -0,0 +1,8 @@
+---
+'@commercetools-test-data/cart-discount-value-gift-line-item': minor
+'@commercetools-test-data/cart-discount': minor
+'@commercetools-test-data/discount-code': minor
+'@commercetools-test-data/commons': minor
+---
+
+Create cart discount presets, empty preset for reference type, empty preset for giftLineItem
```
Comment (stephsprinkle) [inline .changeset/cyan-seals-begin.md:8]:
```suggestion
---

Create cart discount presets, empty preset for giftLineItem
```

**Comment by jaikamat** [comment]
@stephsprinkle The results of removing `Reference.key` 950c09e42d26daecbc094086989a4206d8c88e0d

File: .changeset/cyan-seals-begin.md (hunk 11 above)
Comment (jaikamat) [reply .changeset/cyan-seals-begin.md]: Thanks a bunch! 098e94d6498d68459c4b4d6c83fe9a34ebcd99c2

**Review by stephsprinkle — APPROVED** [review]

**Review by tylermorrisford — APPROVED** [review]
LGTM!

**Pull request merged**